### ISSUE TITLE
latest vsock stable fixes in upstream kernel result in kata failures

### ISSUE THREAD

**Issue by eadamsintel** [kata-containers/runtime#1378]
# Description of problem

I have 3 Clear Linux systems that I let auto update. I noticed late last week that 2 of them failed to run a kata container after updating. Today my personal system also failed to run a kata container.  Error message is "docker: Error response from daemon: OCI runtime create failed: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing: unknown."

# Expected result

The container would be created

# Actual result

OCI error message

---

I tried to run the script at https://github.com/kata-containers/runtime/blob/master/data/kata-collect-data.sh.in  but it said  it couldn't find the runtime. I put the output of kata-env instead but this happens across 3 different systems only after Clear Linux updated. 

$ sudo ./katacollect.sh
Password:
ERROR: katacollect.sh: cannot find runtime '@RUNTIME_NAME@'

```
$ sudo kata-runtime kata-env
[Meta]
  Version = "1.0.20"
[Runtime]
  Debug = false
  Trace = false
  DisableGuestSeccomp = true
  DisableNewNetNs = false
  Path = "/usr/bin/kata-runtime"
  [Runtime.Version]
    Semver = "1.5.0"
    Commit = ""
    OCI = "1.0.1-dev"
  [Runtime.Config]
    Path = "/usr/share/defaults/kata-containers/configuration-qemu.toml"

[Hypervisor]
  MachineType = "pc"
  Version = "QEMU emulator version 2.11.0\nCopyright (c) 2003-2017 Fabrice Bellard and the QEMU Project developers"
  Path = "/usr/bin/kata-qemu-lite-system-x86_64"
  BlockDeviceDriver = "virtio-scsi"
  EntropySource = "/dev/urandom"
  Msize9p = 8192
  MemorySlots = 10
  Debug = false
  UseVSock = false

[Image]
  Path = "/usr/share/kata-containers/kata-containers-image_clearlinux_1.5.0_agent_a581aebf473.img"

[Kernel]
  Path = "/usr/share/kata-containers/vmlinuz-4.14.106-14.container"
  Parameters = "init=/usr/lib/systemd/systemd systemd.unit=kata-containers.target systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket"

[Initrd]
  Path = ""

[Proxy]
  Type = "kataProxy"
  Version = "kata-proxy version 1.5.0"
  Path = "/usr/libexec/kata-containers/kata-proxy"
  Debug = false

[Shim]
  Type = "kataShim"
  Version = "kata-shim version 1.5.0"
  Path = "/usr/libexec/kata-containers/kata-shim"
  Debug = false

[Agent]
  Type = "kata"

[Host]
  Kernel = "5.0.2-717.native"
  Architecture = "amd64"
  VMContainerCapable = true
  SupportVSocks = false
  [Host.Distro]
    Name = "Clear Linux OS"
    Version = "28310"
  [Host.CPU]
    Vendor = "GenuineIntel"
    Model = "Intel(R) Core(TM) i7-6770HQ CPU @ 2.60GHz"

[Netmon]
  Version = "kata-netmon version 1.5.0"
  Path = "/usr/libexec/kata-containers/kata-netmon"
  Debug = false
  Enable = false
```

**Comment by jodh-intel** [comment]
Hi @eadamsintel - thanks for reporting.

Hmm, either the problem is occurring randomly or it might imply we need additional Kata CI tests in Clear?

By the way, that file isn't designed to be run - the `.in` file is the template. When you build the runtime, that template gets converted into a valid shell script called `kata-collect-data.sh`. And that script is supposed to be packaged with the other Kata binaries.

If you happen to know the release when this stopped working, that would help to narrow the issue down.

One final point - Clear isn't using the latest version of Kata (see https://github.com/clearlinux-pkgs/kata-runtime/blob/master/Makefile).

/cc @bryteise.

Related: https://github.com/kata-containers/runtime/issues/1370.

**Comment by eadamsintel** [comment]
Using the Clearlinux release notes as a guide I downgraded to 28240 which works. When I upgrade to 28310 it quits working. According to the release notes the only change is the kata kernel went from 4.14.105-13  -> 4.14.106-14 but the runtime is still 1.5.0

**Comment by eadamsintel** [comment]
As an FYI I downgraded to 28300. Kata works ok.  When upgrading back to version 28310 Kata breaks.

**Comment by jodh-intel** [comment]
Any thoughts on this @bryteise?

**Comment by bryteise** [comment]
I'm a little confused as to how a patch version change to an LTS kernel is causing a breakage. I have updated Kata which should make it to the afternoon build but it seems odd that we didn't see this crash on our end.

**Comment by jodh-intel** [comment]
Ditto. I'm also confused by https://github.com/clearlinux-pkgs/linux-kata/blob/master/Makefile since the version in there is `4.14.90` (for the tip)... and yet we're talking here about *older* releases using `4.14.105` and `4.14.106`...?

I think we need more info on what changed kernel-wise and where those changes are (`.config`). Of course other things will have changed between these two Clear Linux releases including the host kernel (`5.0.1-312 -> 5.0.2-313`). Is it possible to downgrade the host kernel to rule that out?

We probably want input from @jcvenegas and @devimc but they are not around today.

Any thoughts @amshinde, @sboeuf?

I've also raised https://github.com/kata-containers/ci/issues/132.

**Comment by bryteise** [comment]
It doesn't appear to be related to the host kernel (I tried to downgrade there and saw the same issue).

@jodh-intel The Makefile in that case isn't used for the URL and has gotten out of date, if you look at the spec file you'll see the use of 106 as the kernel sources.

**Comment by sboeuf** [comment]
@jodh-intel nothing from the top of my head.

/cc @mcastelino as he's using CL with Kata on a regular basis I believe

**Comment by bryteise** [comment]
The only change between 105 and 106 is whatever changed in the kernel sources, no config, the rest of the changes between versions don't have any obvious problem children. May need to bisect between 105 and 106 but first we can try running 28300 but sneaking in the 28310 kata kernel.

**Comment by bryteise** [comment]
So I have tried using the 105 kernel in 28310 and the bug doesn't show. Using the 106 kernel in 28300 has the bug so the issue definitely seems to stem from a change in the 4.14 LTS between 105 and 106.

**Comment by jodh-intel** [comment]
@bryteise - thanks for looking into this! Could you possibly post a kernel config diff so we can compare that to the known good configs in?:

- https://github.com/kata-containers/packaging/tree/master/kernel/configs

**Comment by bryteise** [comment]
I tracked it down to a specific commit related to vsocks but I see 107 just got released with some vsock changes so I'm going to try that and see if it fixes. If that doesn't work then I'll post hopefully useful information =).

**Comment by grahamwhaley** [comment]
thx @bryteise - if you happen to have a ref to the commit(s), we have a vsocks person here on the kata repos. /cc @stefanha

**Comment by bryteise** [comment]
[here](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-4.14.y&id=6a78f5dccfff5c0b64b6956e6f6beadd669e0108) is the breaking commit for me. 107 didn't fix the issue for me unfortunately. The diff between our config and your 4.14 config is [here](https://gist.github.com/bryteise/62c42d6e32141cf325de914687df6664).

**Comment by sboeuf** [comment]
We might need to fix something in Kata because of this patch. But I agree having feedback from @stefanha would be great here.

**Comment by stefanha** [comment]
@bryteise Please try applying this patch:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/net/vmw_vsock?id=4c404ce23358d5d8fbdeb7a6021a9b33d3c3c167

**Comment by bryteise** [comment]
@stefanha After applying the patch I still see the same failure to start.

**Comment by sboeuf** [comment]
Any chance Kata Containers (runtime or agent) needs to be updated (I mean with a proper patch) regarding the recent kernel patches related to vsock?

**Comment by stefanha** [comment]
> @stefanha After applying the patch I still see the same failure to start.

Please check the Kata sandbox's startup messages for errors or backtraces.

You can do this by editing /etc/kata-containers/configuration.toml:
[hypervisor.qemu]
path = "/path/to/your/kata-qemu-wrapper"
kernel_params = "... console=ttyS0"

The wrapper shell script looks like this:
$ cat kata-qemu-wrapper
#!/bin/bash
exec /usr/bin/qemu-system-x86_64 "$@" -serial file:/tmp/serial.log

After launching a Kata container fails, check /tmp/serial.log on the host.

**Comment by bryteise** [comment]
@stefanha Well after prodding I updated the kernel to 4.19.31 and reapplied the patch, before and after produces the same log file though:
```
[    0.228447] EXT4-fs (pmem0p1): DAX unsupported by block device. Turning off DAX.                                                               [0/0]
[    0.280280] systemd[80]: tmp.mount: Failed to execute command: No such file or directory                                                           
time="2019-03-26T22:45:40.874116778Z" level=info msg=announce device-handlers="nvdimm,mmioblk,blk,scsi" name=kata-agent pid=109 source=agent storage-he
time="2019-03-26T22:45:40.877120902Z" level=info msg="Started listening for uevents" name=kata-agent pid=109 source=agent subsystem=udevlistener      
time="2019-03-26T22:45:40.878143818Z" level=info msg="agent grpc server starts" name=kata-agent pid=109 source=agent 
```

**Comment by sboeuf** [comment]
@bryteise ohh wait you need to use the latest image here. If you don't have the metadata written to the image between 4-8k, you will not get NVDIMM detected properly and there is no way the rootfs will be booted properly.

**Comment by bryteise** [comment]
@sboeuf Alrighty, I'll try again with tomorrows build using Kata 1.6.0.

**Comment by amshinde** [comment]
@bryteise So I tried this  on a clearlinux VM under vagrant with the clear-linux kernel as
```
uname -r
4.19.31-16.lts2018
```
The kata-kernel is vmlinuz-4.14.108-19.container.
I was able to have Kata working with use_vsock=true set in the confiuration.toml
Kata did not work with use_vsock=false which is the default.

I know you havent seen the same on bare-metal. I'll try to reproduce this once 1.6.0 is out tomorrow.

**Comment by bryteise** [comment]
Okay, using Kata 1.6.0 I'm seeing:

```
[    0.359886] EXT4-fs (pmem0p1): DAX unsupported by block device. Turning off DAX.
[    0.477009] systemd[1]: systemd-udevd-kernel.socket: Socket service systemd-udevd.service not loaded, refusing.
[    0.500213] systemd[1]: Failed to listen on udev Kernel Socket.
time="2019-03-27T21:47:25.348137338Z" level=info msg=announce device-handlers="mmioblk,blk,scsi,nvdimm" name=kata-agent pid=85 source=agent storage-ha4
time="2019-03-27T21:47:25.357191965Z" level=info msg="Started listening for uevents" name=kata-agent pid=85 source=agent subsystem=udevlistener
time="2019-03-27T21:47:25.377778752Z" level=info msg="agent grpc server starts" name=kata-agent pid=85 source=agent
time="2019-03-27T21:47:25.381743014Z" level=info msg="Waiting for stopServer signal..." name=kata-agent pid=85 source=agent subsystem=stopserverwatcher
```

**Comment by jcvenegas** [comment]
I think `DAX unsupported by block device. Turning off DAX.` is expected. Does it boot now or you see any other error output ?

**Comment by jodh-intel** [comment]
@bryteise - could you enable full debug so we can get more logs?:

- https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#enable-full-debug

**Comment by devimc** [comment]
@bryteise to fix the DAX issue you need to generate a new rootfs image
see https://github.com/kata-containers/osbuilder/pull/264

**Comment by sboeuf** [comment]
@devimc is the updated image not part of our 1.6 packages?

**Comment by devimc** [comment]
@sboeuf  we haven't released 1.6.0, so I guess it can be part of it

**Comment by devimc** [comment]
@sboeuf ohh sorry, I didn't see the release, no it's not part of that release,

**Comment by sboeuf** [comment]
@jcvenegas @devimc let's fix it :)

**Comment by devimc** [comment]
@sboeuf  1.6.1 ?

**Comment by sboeuf** [comment]
@devimc oh you mean the patches were not part of the 1.6.0 release itself?

Anyway, not a huge deal, we're only talking about the DAX enhancement, nothing functional being broken here.

**Comment by devimc** [comment]
@sboeuf yes, nothing got broken, we can fix the mem footprint in a next release 1.6.1 or 1.7.0

**Comment by amshinde** [comment]
@bryteise I tried this on an Ubuntu setup with the Clearlinux kernels. I saw the same issues as outlined here. But, I did get Kata working with `use_vhost=true` set in the configuration.toml.
I did have to modprobe vhost-vsock before thta.
Some debugging today showed that 4.14.106 kernel was introducing issues on virtio-serial channel that we start the agent grpc channel on(the listen() call  on the channel is stuck) Will debug some more to find the exact cause for this.

**Comment by amshinde** [comment]
@bryteise Meanwhile I just tried this on a Clearlinux box with use_vhost=true and vhost-vsock module loaded. Kata works in that case. So to get Kata  working in the meantime on Clear, we can add a patch to enable that config. We should really be setting that as the default, but this is something that will have to wait till the next Kata release.
cc @sboeuf

**Comment by sboeuf** [comment]
@stefanha 
Ok so @amshinde and I spent some time debugging this today, and the good news is that we managed to spot the issue, but we need some feedback to understand if this needs to be fixed in Kata Containers or directly in the kernel.

What happens is that this [patch](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-4.14.y&id=6a78f5dccfff5c0b64b6956e6f6beadd669e0108) introduces a big change regarding the life cycle of the `vsock` guest driver. Instead of having registration of the `AF_VSOCK` type in the probing function, it is moved to the driver initialization function. In the case that is failing here, we are using Kata __without__ using `vhost-vsock-pci`, which does not trigger the probe of the driver. Because the driver is _built-in_, the initialization happens no matter if the VM exposes a `vsock` device or not.
The problem is, our agent has a mechanism that will check if vsock is supported by the guest, and will assume about the socket type based on this. The way we check for vsock happens in two steps. First the agent checks for the path `/dev/vsock` existence, and if it's found, as a second step, it will issue the following call:
```go
fd, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
```
to check if `AF_VSOCK` is __really__ supported.
Problem is, in this particular case, the agent will think that `vsock` is fully supported, hence it will start opening an `AF_VSOCK` socket and it will start listening but nothing will ever come... 

Now, I'm not saying this cannot be fixed in the Kata agent, we could always find a way, but I'm wondering about the validity of the `vsock` driver behavior here. By moving `vsock_core_init()` to the init function `virtio_vsock_init()`, the `AF_VSOCK` socket type now gets registered, but is clearly unusable since there is not a single device backing it. It feels to me that `vsock_core_init()` should be moved back the probe function `virtio_vsock_probe()` to behave properly. But I'm not an expert in this field, and having the socket type registered independently from the fact that there is a device backing it might be the expected behavior.

**Comment by stefanha** [comment]
> @stefanha
> Ok so @amshinde and I spent some time debugging this today, and the good news is that we managed to spot the issue, but we need some feedback to understand if this needs to be fixed in Kata Containers or directly in the kernel.
> 
> What happens is that this [patch](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=linux-4.14.y&id=6a78f5dccfff5c0b64b6956e6f6beadd669e0108) introduces a big change regarding the life cycle of the `vsock` guest driver. Instead of having registration of the `AF_VSOCK` type in the probing function, it is moved to the driver initialization function. In the case that is failing here, we are using Kata **without** using `vhost-vsock-pci`, which does not trigger the probe of the driver. Because the driver is _built-in_, the initialization happens no matter if the VM exposes a `vsock` device or not.
> The problem is, our agent has a mechanism that will check if vsock is supported by the guest, and will assume about the socket type based on this. The way we check for vsock happens in two steps. First the agent checks for the path `/dev/vsock` existence, and if it's found, as a second step, it will issue the following call:
> 
> ```go
> fd, err := unix.Socket(unix.AF_VSOCK, unix.SOCK_STREAM, 0)
> ```
> 
> to check if `AF_VSOCK` is **really** supported.
> Problem is, in this particular case, the agent will think that `vsock` is fully supported, hence it will start opening an `AF_VSOCK` socket and it will start listening but nothing will ever come...
> 
> Now, I'm not saying this cannot be fixed in the Kata agent, we could always find a way, but I'm wondering about the validity of the `vsock` driver behavior here. By moving `vsock_core_init()` to the init function `virtio_vsock_init()`, the `AF_VSOCK` socket type now gets registered, but is clearly unusable since there is not a single device backing it. It feels to me that `vsock_core_init()` should be moved back the probe function `virtio_vsock_probe()` to behave properly. But I'm not an expert in this field, and having the socket type registered independently from the fact that there is a device backing it might be the expected behavior.

The VMware VMCI vsock driver behaves in the same way.  Vsock doesn't support probing the availability of a transport.

kata-agent is relying on side-effects of the vmw_vsock_virtio_transport.ko driver, but it's not a supported interface.

Two alternatives that are still virtio-specific and not guaranteed to be stable:

1. Open /dev/vsock and invoke IOCTL_VM_SOCKETS_GET_LOCAL_CID.  When there is no virtio transport available the CID is VMADDR_CID_ANY (0xffffffff).  Note that this is still virtio-specific because the VMware VMCI vsock driver will return VMADDR_CID_HOST (0x2) instead.

2. Check if the vmw_virtio_virtio_transport.ko driver has probed a virtio device:
```
$ ls -l /sys/bus/virtio/drivers/vmw_vsock_virtio_transport
...
lrwxrwxrwx. 1 root root    0 Mar 29 17:54 virtio0 -> ../../../../devices/pci0000:00/0000:00:04.0/virtio0
```
Or check that /proc/bus/pci/devices has a 1af41012 (virtio-vsock) device.

**Comment by stefanha** [comment]
Out of the two alternative probing approaches I would recommend checking for the PCI device.  The IOCTL_VM_SOCKETS_GET_LOCAL_CID approach won't work on modular kernels because they are likely to load the VMware VMCI vsock driver if the virtio device is not available.

**Comment by amshinde** [comment]
I was leaning towards the second approach that @stefanha listed, but taking a step back I think we can simplify the whole flow by simply having the runtime pass the information about vsock is being used or not to the agent using a kernel option similar to how we do this currently for `agent.debug`.
That way the agent gets full knowledge about what communication mechanism to use at startup rather than detecting it later.
So this could be a kernel option like `agent.use_vsock=true/false`.

cc @sboeuf

**Comment by amshinde** [comment]
cc @devimc

**Comment by egernst** [comment]
This should target 1.6.x, 1.5.x and 1.7-alpha release, due next week.  

In the mean time, work around is to utilize vsock if you are using a guest kernel which exhibits this failure.

**Comment by sboeuf** [comment]
Thanks @stefanha for the feedback, and thanks @bryteise for spending some time digging, I think we now have a good plan to get this addressed.

**Comment by mdlayher** [comment]
:wave: I'm a little late here, but I'd be willing to add a bit of code to github.com/mdlayher/vsock to probe for the existence of the virtio-vsock device if that's what's truly required to make vsock usable.

I admit, I am confused about the differences between the VMWare and virtio device implementations. I've only ever tried my package with the virtio device and actually have no idea if it'd work for the VMWare version too.

**Comment by sboeuf** [comment]
@mdlayher 
> I'd be willing to add a bit of code to github.com/mdlayher/vsock to probe for the existence of the virtio-vsock device if that's what's truly required to make vsock usable.

The problem by adding this as part of your vsock package is that you would have to come up with a global solution that works for all cases, and you cannot afford to make some assumptions.
From our agent perspective, we can actually assume that we are not running in the vmware context, hence we can do a quick assumption about checking `virtio-vsock` device.
Actually, you could maybe check for either a `virtio-vsock` or vmci device presence, I'm not sure.

>I admit, I am confused about the differences between the VMWare and virtio device implementations. I've only ever tried my package with the virtio device and actually have no idea if it'd work for the VMWare version too.

Now if you're clearly stating that this package works only for virtio-vsock, it would be easier to integrate the vsock presence as part of your code.

**Comment by mdlayher** [comment]
> The problem by adding this as part of your vsock package is that you would have to come up with a global solution that works for all cases, and you cannot afford to make some assumptions.

Seems reasonable to me to leave that check in my consumer's code for the time being then. If you end up creating a general solution that every consumer of my package will need, I'm happy to accept a patch though.

> Now if you're clearly stating that this package works only for virtio-vsock, it would be easier to integrate the vsock presence as part of your code.

That's the thing: I'm not actually sure! I've never tried it with the VMCI device and don't have access to a system where I could try that out. I have always developed against the virtio-vsock device.

**Comment by TerrenceXu** [comment]
I also reproduced this issue.
When I install use clearlinux: 28490 with kata 1.5.1, I run container with kata runtime and printed "docker: Error response from daemon: OCI runtime create failed: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing: unknown. "  
When I downgraded clearlinux to 28240 with kata 1.5.0, this issue gone, container run well with kata runtime.

**Comment by grahamwhaley** [comment]
@bryteise - did this get tracked down and fixed yet?

**Comment by bryteise** [comment]
I'm not sure, it seems like this was fixed via #1458 though I haven't checked if that is in a release as I've been out of it for two weeks now X(.

**Comment by TerrenceXu** [comment]
I just checked the latest clearlinux version (28750) with Kata 1.6.0 and I still can meet this issue. :(

**Comment by sboeuf** [comment]
The fix is not part of the 1.6 release.
/cc @amshinde

**Comment by amshinde** [comment]
This went into the 1.6.1 release. I had spoken to @gtkramer about this yesterday, he mentioned the fix would go into today mornings release. Will let you know once I can confirm this.

**Comment by TerrenceXu** [comment]
Thanks @amshinde, the clearlinux version (28780) with Kata 1.6.1 is working. :)

**Comment by amshinde** [comment]
Thanks @TerrenceXu for verifying :). Closing this issue!
